Add CircleCI build status widget

monfresh opened the pull request:
The Circle CI API does not currently provide a way to get a list of just
the active repos, so we have to use the GitHub API to get a list of all
repos that belong to an organization, then make a separate Circle API call for each one (over 300 for 18F), just to determine whether Circle CI has builds for that repo.

In order to minimize the amount of times we have to do this, I created a new job that runs once a day and stores this list in an env var, which can then be reused by the Circle CI widget, which runs every 5 minutes.

I also did the same optimization for the list of GitHub org repos.

To support this new optimization, I adjusted the start time of each job so that it would allow enough time to pass for the env vars to get populated.

Closes #14.

monfresh committed:
Add CircleCI build status widget

The Circle CI API does not currently provide a way to get a list of just
the active repos, so we have to use the GitHub API to get a list of all
repos that belong to an organization, then make a separate Circle API
call for each one (over 300 for 18F), just to determine whether
Circle CI has builds for that repo.

In order to minimize the amount of times we have to do this, I created
a new job that runs once a day and stores this list in an env var,
which can then be reused by the Circle CI widget, which runs every 5
minutes.

I also did the same optimization for the list of GitHub org repos.

To support this new optimization, I adjusted the start time of each
job so that it would allow enough time to pass for the env vars to get
populated.

Closes #14.

monfresh commented:
Oops, I forgot to run the tests in my excitement. I'll fix them and add new ones for the new feature.

monfresh commented:
Alrighty. Tests have been written and we are green all around. Ready for review!
cc @jessieay @mbland @arowla

mbland commented on jobs/active_circle_repos.rb:
```diff
@@ -0,0 +1,21 @@
+# the active repos, so we have to use the GitHub API to get a list of all
+# repos that belong to an organization, then make a separate Circle API call
+# for each one (over 300 for 18F), just to determine whether Circle CI has
+# builds for that repo.
```
Ugh.

arowla replied:
Can we use the Travis API to subtract those from the list of all repos to try on CircleCI?

monfresh replied:
That could be another option, but I think what we have now is good enough. Eventually, once all repos have an `.about.yml`, we'll use the Team API to store a list of testable repos and only query those.

mbland commented on jobs/github_org_repos.rb:
```diff
@@ -0,0 +1,11 @@
+
+# This job runs once a day to store the list of all GitHub repos for an org
+# in an env var. That way, it can be reused by other widgets without having
+# to make more API calls.
```
It'd be nice if there was a way to trigger this via an org webhook, to add new repos as they happen to appear, and remove old ones if they go away. Obviously a future enhancement, but do you know if that's possible?

monfresh replied:
Not off the top of my head.

mbland commented on spec/jobs/active_circle_repos_spec.rb:
```diff
@@ -0,0 +1,41 @@
+
+  it 'calls CircleBackend.get_last_build_for' do
+    expect(CircleBackend).
+      to receive(:get_last_build_for).ordered.with('github-dashing').and_return({})
```
This isn't supposed to read `to.receive`?

monfresh replied:
Nope. It's `.to receive`, just like all other `.to` expectations, such as `.to eq`. If the syntax were wrong, the test would have failed I presume.

mbland commented on spec/jobs/active_circle_repos_spec.rb:
```diff
@@ -0,0 +1,41 @@
+  it 'sets ACTIVE_CIRCLE_REPOS env var to active repos only' do
+    job.trigger_block
+
+    expect(ENV['ACTIVE_CIRCLE_REPOS']).to eq 'save-ferris'
```
Should probably clear this var in the before/after setup.

monfresh replied:
Good point.

jessieay replied:
alternatively, you could stub `ENV`:

`allow(ENV).to receive(:[]).with("ACTIVE_CIRCLE_REPOS").and_return("save-ferris")`

monfresh replied:
In this case, running the test actually creates the ENV var, so we want to test that it got created, but then clear it out so that it's empty for the next test.

jessieay commented on jobs/active_circle_repos.rb:
```diff
@@ -0,0 +1,21 @@
+# repos that belong to an organization, then make a separate Circle API call
+# for each one (over 300 for 18F), just to determine whether Circle CI has
+# builds for that repo.
+SCHEDULER.every '1d', first_in: '50s' do |_job|
```
NICE - is this deployed to CF?I need to schedule daily cron jobs and someone suggested Clockwork, which seems like a decent option. Will look into how you've set this up if it is on CF...

monfresh replied:
It is on CF, and I didn't have to do anything special. I just pushed the app to CF and set the appropriate env vars and that was it. [Dashing](https://github.com/Shopify/dashing) (the framework this app is built on) uses [rufus-scheduler](https://github.com/jmettraux/rufus-scheduler).

Most of the hard work has already been done by Dashing, including providing all kinds of widget types. You can also browse other widgets people have built: https://github.com/Shopify/dashing/wiki/Additional-Widgets

jessieay commented on jobs/active_circle_repos.rb:
```diff
@@ -0,0 +1,21 @@
+SCHEDULER.every '1d', first_in: '50s' do |_job|
+  org_repos = ENV['GITHUB_ORG_REPOS'].split(',')
+
+  circle_builds = org_repos.map { |repo| CircleBackend.get_last_build_for(repo) }.flatten.compact
```
try `flat_map` http://ruby-doc.org/core-2.1.0/Enumerable.html#method-i-flat_map

monfresh replied:
Yes! I need to remember to run fasterer to catch these things: https://github.com/DamirSvrtan/fasterer

jessieay commented on spec/support/fake_circle_api.rb:
```diff
@@ -0,0 +1,13 @@
+include JobHelper
+
+class FakeCircleApi < Sinatra::Base
+  get '/api/v1/project/18F/github-dashing' do
```
YES love this

jessieay commented:
Couple of comments from me, but looks good to merge when you're ready.

mbland commented on spec/jobs/github_org_repos_spec.rb:
```diff
@@ -0,0 +1,37 @@
+  it 'sets GITHUB_ORG_REPOS env var to all org repos' do
+    job.trigger_block
+
+    expect(ENV['GITHUB_ORG_REPOS']).to eq 'github-dashing,save-ferris'
```
Clear the env var in before/after.

monfresh committed:
Reduce duplication in specs and improve setup

- Clear out env vars before and after each test
- Create job_matching method to reduce duplication
- Use flat_map for faster performance

monfresh commented:
Thanks for the reviews, everyone! I added a new commit to address your comments, and also took care of some duplication in the tests. Let me know what you think.

jessieay commented:
I left a comment before saying "left a few comments but looks good to merge when ready" and I stand by it!
